### DISCUSSION TITLE
[Question] Is LED 1.29 incompatible to IPED?

### DISCUSSION THREAD

**Discussion by thalespr** [sepinf-inc/IPED#706] [Q&A] [answered]
Not sure if it was dealed in another issue, but I want to know if the latest version of LED (1.29) is incompatible to IPED? Tried to run some processing using it, but got some errors. If it is, I would like to ask to gently leave an issue about that opened (if there isn't one already...) in order to keep track on this and other people be aware of that.

**Comment by wladimirleite** [comment]
In fact it is intentionally incompatible, so I guess it is not a issue.
It is explicitly stated on LED's "LEIAME.txt" notes of 1.29.00 release, LED internal WIKI (versions section) and I warned about that in the e-mail in which LED 1.29 release was announced, but I know people usually don't read any of these :-)

The new IPED version (4.x) won't be compatible as it was before (pointing to LED's folder), but I will release a CSV file with LED's database to be used with IPED.

That being said, I am fine with keeping this issue open until IPED 4.x is released, but I am not sure how helpful it would be.
Not sure, but maybe moving to discussion would make more sense.

**Comment by thalespr** [comment]
> In fact it is intentionally incompatible, so I guess it is not a issue.
> It is explicitly stated on LED's "LEIAME.txt" notes of 1.29.00 release, LED internal WIKI (versions section) and I warned about that in the e-mail in which LED 1.29 release was announced, but I know people usually don't read any of these :-)
> 
> The new IPED version (4.x) won't be compatible as it was before (pointing to LED's folder), but I will release a CSV file with LED's database to be used with IPED.
> 
> That being said, I am fine with keeping this issue open until IPED 4.x is released, but I am not sure how helpful it would be.
> Not sure, but maybe moving to discussion would make more sense.

 I honestly don't remember reading email about that (If I did, I couldn't recall that). So I spent some time to discover that it wasn't compatible to IPED on purpose (shame on me I didn't look up for LED Releases Notes, my bad for sure). At first sight, looking IPED output, I thought it could be a bug. Maybe other people can think the same, that's why I suggested to keep an issue opened.

You said that IPED 4.x won't be compatible? Is there a plan to make it compatible?

Sharing CSV file along LED database will be appreciated since IPED is able to import it AFAIK.

**Reply by lfcnassif** [reply]
> Sharing CSV file along LED database will be appreciated since IPED is able to import it AFAIK.

It's not, but 4.0 will be.

**Comment by wladimirleite** [comment]
> You said that IPED 4.x won't be compatible? Is there a plan to make it compatible?

The current compatibility is convenient in the sense that users of both tools can just point to LED's database when setting IPED.
But there are a  few drawbacks: 
- I can't change LED's format, which I sometimes need/want to do, without changing IPED.
- There are information used by LED that are useless to IPED, so IPED needs to read a larger file.
- The opposite also happens (e.g. PhotoDNA of LED database files can be used in IPED, but not in LED). 

The idea for the new IPED version is to have a simpler and unique way to handle all hash sets (LED, NSRL, ProjectVIC, PhotoDNA, and possibly user custom hash sets).
So LED database will be just another hash database, in a standard CSV format.
Whenever I update LED database, I will also release (in our internal server) a LED CSV (and a SQLite with "everything" already imported).
In the future, there will be a centralized CSAM database, and probably a CSV release from time to time.

By the way, I plan to release LED 1.29.01 this week.
I will try to highlight this limitation, as I totally agree it is confusing, while IPED 4.x is not released.

**Comment by lfcnassif** [comment, marked as answer]
@thalespr see #118 for the new centralized hash database details. I agree and will move this to discussions. As IPED-3.18.8 is around, I will add a warning to state that LED >= 1.29 format is not compatible anymore.

**Comment by thalespr** [comment]
Thanks @tc-wleite and @lfcnassif!

**Comment by vtgdias** [comment]
@tc-wleite, I am using LED 1.32, is there any script to convert TXT to CSV format, so i can add LEDs database to iped hashset database?

**Comment by wladimirleite** [comment]
Hi @vtgdias!

I have the following Java code.
You need compile it and run passing 2 parameters: the input folder and the output file, like:
`java LedHashesToCSV D:\LED-1.32\hashes-txt D:\IPED\hashes\led-1.32.csv`

```java
import java.io.*;
import java.util.*;

public class LedHashesToCSV {
    public static void main(String[] args) throws Exception {
        if (args.length != 2) return;
        String str = null;
        File[] files = new File(args[0]).listFiles();
        List<String> lines = new ArrayList<String>();
        for (File f : files) {
            String grp = f.getName();
            int p = grp.lastIndexOf('.');
            if (p > 0) grp = grp.substring(0, p);
            int n = 0;
            BufferedReader in = new BufferedReader(new InputStreamReader(new FileInputStream(f)));
            StringBuilder sb = new StringBuilder();
            while ((str = in.readLine()) != null) {
                if (str.charAt(0) == '#') continue;
                n++;
                String[] s = str.trim().split(" \\*");
                if (s.length < 7) {
                    System.err.println("Invalid format in file " + f.getName() + ", line " + n);
                    in.close();
                    return;
                }
                String ext = s[9];
                int q = ext.lastIndexOf('.');
                if (q >= 0) ext = ext.substring(q + 1).toLowerCase();
                else ext = "";
                if (ext.length() > 16) ext = "";
                String cat = s[8];
                sb.append(s[0].toUpperCase()).append(",");
                sb.append(s[3].toUpperCase()).append(",");
                sb.append(s[2].toUpperCase()).append(",");
                sb.append(s[6].toUpperCase()).append(",");
                sb.append(cat.equals("1") ? "pedo" : "").append(",");
                sb.append("LED").append(",");
                sb.append(s[4].toUpperCase()).append(",");
                sb.append(s[1].toUpperCase()).append(",");
                sb.append(String.valueOf(Long.parseLong(s[7]))).append(",");
                sb.append(ext).append(",");
                sb.append(grp).append(",");
                sb.append(cat);
                lines.add(sb.toString());
                sb.delete(0, sb.length());
            }
            in.close();
            System.err.println(f.getName() + " : " + n);
        }
        BufferedWriter out = new BufferedWriter(new FileWriter(args[1]), 1 << 20);
        out.write("MD5,SHA1,EDONKEY,SHA256,status,set,md5_512,md5_64k,fileLength,fileExt,ledGroup,ledCategory");
        out.newLine();
        Collections.sort(lines);
        for (String s : lines) {
            out.write(s);
            out.newLine();
        }
        out.close();
    }
} 
```

**Reply by vtgdias** [reply]
Tks @tc-wleite
